File: packages/restapi/src/lib/user/getUsersBatch.ts
import axios from 'axios';
import { IUser } from '../types';
import { isValidETHAddress, walletToPCAIP10 } from '../helpers/address';
import { getAPIBaseUrls, verifyPGPPublicKey } from '../helpers';
import Constants, { ENV } from '../constants';

export interface GetBatchType {
  userIds: string[];
  env?: ENV;
}

export const getBatch = async (options: GetBatchType): Promise<IUser> => {
  const { env = Constants.ENV.PROD, userIds } = options || {};

  const API_BASE_URL = getAPIBaseUrls(env);
  const requestUrl = `${API_BASE_URL}/v1/users/batch`;

  const MAX_USER_IDS_LENGTH = 100;
  if (userIds.length > MAX_USER_IDS_LENGTH) {
    throw new Error(
      `Too many user IDs. Maximum allowed: ${MAX_USER_IDS_LENGTH}`
    );
  }

  for (let i = 0; i < userIds.length; i++) {
    if (!isValidETHAddress(userIds[i])) {
      throw new Error(`Invalid user address!`);
    }
  }

  const pcaipUserIds = userIds.map(walletToPCAIP10);
  const requestBody = { userIds: pcaipUserIds };

  return axios
    .post(requestUrl, requestBody)
    .then((response) => {
      response.data.users.forEach((user: any, index: number) => {
        response.data.users[index].publicKey = verifyPGPPublicKey(
          user.encryptionType,
          user.publicKey,
          user.did,
          user.nftOwner
        );
      });
      return response.data;
    })
    .catch((err) => {
      console.error(`[Push SDK] - API ${requestUrl}: `, err);
      throw Error(`[Push SDK] - API ${requestUrl}: ${err}`);
    });
};

File: packages/push-api-dart/lib/src/user/src/get_users_batch.dart
import '../../../push_api_dart.dart';

getBatch({
  required List<String> userIds,
}) async {
  final maxUserIdsLength = 100;
  if (userIds.length > maxUserIdsLength) {
    throw Exception('Too many user IDs. Maximum allowed: $maxUserIdsLength');
  }

  for (int i = 0; i < userIds.length; i++) {
    if (!isValidETHAddress(userIds[i])) {
      throw Exception('Invalid user address!');
    }
  }

  final pcaipUserIds = userIds.map(walletToPCAIP10);
  final requestBody = {'userIds': pcaipUserIds};

  final result = await http.post(
    path: '/v2/users/batch',
    data: requestBody,
  );

  if (result == null) {
    return null;
  }

  if (result['users'] is List) {
    final output = [];
    for (var item in result['users']) {
      var user = User.fromJson(item);
      user.publicKey = verifyPGPPublicKey(
        encryptedPrivateKey: user.encryptedPrivateKey!,
        publicKey: user.publicKey!,
        did: user.did!,
      );

      user = await populateDeprecatedUser(user: user);
      output.add(user);
    }

    
  }
}
